### DISCUSSION TITLE
Ecosystem visibility of a Filecoin (subnet) consensus with Narwhal+Tusk behind the scenes

### DISCUSSION THREAD

**Discussion by aakoshh** [consensus-shipyard/consensuslab#165] [General]
# Problem Statement
The purpose of this discussion is to get a common agreement on the following:
* What is the vision for what external users who can only observe Filecoin blocks should expect from a subnet or standalone net running a consensus that uses Narwhal+Task/Bullshark for total ordering of transactions?
* What that means exactly when we look at integrating this into Forest or Lotus?

The premise is that the output blocks will conform to Filecoin's chain structure, which currently (in Forest) the mempool gossiping transactions, mean miners selecting transactions in the correct sequence based on account nonces, creating and signing blocks arranged in tipsets, blocks gossiped around and using consensus specific validation rules before executing the Wasm transactions. 

Once transactions arrive from a black box external total ordering consensus component such as Narwhal+Tusk/Bullshark, there are some open questions regarding the Filecoin blocks created. Should they be signed by a miner? Should they look like longest chain consensus? Should it support existing explorers, maybe light clients, bridges?

# Proposals
I would like to propose two possible approaches to integrating Narwhal into Forest: one that preserves ecosystem support to the fullest, and another which doesn't. 

## Common part

The following would apply to both approaches:
* Disable mempool gossiping completely, deliver locally received user transactions to the Narwhal node after pre-ordering them for account sequences. This achieves parallelism across nodes, and reduces duplication between transaction batches to the extent that users might submit their transactions to multiple nodes for redundancy.
* Receive a steady stream of (potentially empty) committed batches from Tusk/Bullshark. 
* Execute the transactions, ignoring duplicates; issue a penalty for out-of-order messages as those arrived from malicious batches.
* Somehow attribute rewards/penalties to whichever miner batched the transaction. 

### Could Filecoin do this already? 

One tangible benefit over vanilla Filecoin is the parallelisation effect that nodes can achieve by only worrying about the partial order withing the batches they contribute, e.g. they don't have to consider transactions from 10,000 users, just the 100 that deployed to them, which means less database access and cache volume. But this should be possible to achieve with vanilla Filecoin as well because of how Tipsets allow multiple blocks at the same height; all that's needed is to disable the mempool doing its own gossip, no need for Narwhal.

Also note that the Filecoin blocks that get gossiped around do not contain the full transaction payload, only CIDs, there is no double transmission. The only obstacle for parallelisation is if the mempool saturates the network with the same transactions, so everyone proposes the sameish blocks in the tipset. 

Another thing is that at least in Forest, the mempool only gossips locally transmitted transactions, there is no N^2 cost for the broadcast. 

## Option 1: Full ecosystem compatibility

This version applies a (hopefully lightweight) Filecoin longest chain consensus over the batches coming from Tusk:
* Deterministically pick (some) random miner(s) in each epoch eligible to sign Filecoin blocks. Tipsets allow redundancy here.
* If eligible to produce a block in the epoch that comes after the tip of the last Filecoin block, slice the transactions into Filecoin blocks in that order as is even if some messages turn out to be duplicates or out of order.
* Gossip the Filecoin blocks.
* For validation, check that the transactions in the body correspond exactly to the order from Tusk/Bullshark, and that miners were eligible in the epoch.
* If all eligible Filecoin miners in an epoch are absent, the miner can increment the epoch to see if they become eligible to produce a block when the next Tusk block arrives, effectively using some kind of a timeout. But they will always pick transactions from the prefix of the pending queue of committed Tusk batches.
* There is a longest chain fork choice weight function to decide whole fork in Filecoin to follow, which should take into account the transaction volume and potentially the epochs at which they appeared, so blocks that published transactions earlier are better, but doesn't allow a miner to hide their blocks for too long because the overall volume on miners who stepped in will dominate it.
* The historical Filecoin block sync method won't be able to fully verify blocks beyond the garbage collection limit of Narwhal+Tusk, because a latecomer cannot retrieve the original batch stream any more from peers. It has to trust that the longest chain will prevail. It can also get a snapshot. Checkpoints are currently not part of the Filecoin sync protocols, it looks for a sample size of the best headers to decide where to start going back towards its own tip from.

This doesn't even need checkpoints to look good for explorers/users/light clients because it's just longest chain with miners showing their agreement in the form of their signatures in subsequent blocks. Checkpoints will be necessary for the parent subnet, though.

It's not obvious why we would introduce the potential latency of forking on Filecoin over a total ordered transaction list; it most likely defeats all advantages of having the efficient reliable broadcast of Narwhal and communication-free total ordering of Tusk to do so. Once again this is my proposal for fitting in with the existing ecosystem. 

## Option 2: Reduced compatibility

This version doesn't do the longest chain, so it's less complex, but needs changes in the ecosystem:
* Nobody signs the blocks. 
* Blocks are are deterministically generated from Tusk/Bullshark.
* Blocks are not gossiped between nodes, everyone makes them locally.
* For this reason blocks don't need to be validated, all we require is everyone executes them following the same rules.
* At regular intervals the nodes include a special transactions to produce a checkpoint that includes that block hash (not just state hash), and later cross messages etc. Then the validators start sending further special transactions to vote on this checkpoint if they agree, allowing explorers/users to look for them.
* For late joiners, they need a snapshot of the chain and state, because Narwhal+Tusk will garbage collect the old batches and the node can no longer derive the Filecoin blocks. It could potentially use the built-in historical block sync mechanism to fetch Filecoin blocks from others and look for checkpoints in them, but it would perhaps be strange not to gossip blocks, but request their history. Best is to acquire a snapshot within the garbage collection period and let Narwhal sync from there. 

This version has the benefit of avoiding the potential of increased latency due to forking/inactive miners in the Filecoin nodes, and as long as blocks are kept deterministic still should lead to agreement across nodes, as evidenced by the checkpoints that will emerge. It would be up to the community to evolve their tooling to support these checkpoints. 

# Prior art

The following docs already capture most of the internal architecture of using Narwhal+Tusk with Eudico: 
* [Incremental Design of Eudico's Ordering Layer](https://hackmd.io/1vxyguGMSlWZM9_s8PCFbw) 
* [Implementation of Eudico’s Ordering Layer](https://hackmd.io/P59lk4hnSBKN5ki5OblSFg)

They are more complete with regards to configuration changes for example, however, they are do not explicitly say anything about how Eudico blocks are treated, how much of the existing block sync/gossip machinery is still kept in place. They also have a different vision for using the mempool.

**Comment by vukolic** [comment]
As you correctly remark in the last paragraph in option 1 - having Filecoin consensus on top of Narwahl + Tusk/Bullshark "likely defeats all advantages of ... Narwahl and ... Tusk".

The idea for Hiearchical consensus subnets was always to go with an approach which is similar to your Option 2. 

I suggest to have a brief call to sync on different points you raise here, together with @matejpavlovic to discuss the approach Mir project (https://github.com/filecoin-project/mir) takes in Eudico, which boils down to same considerations. We can then report here to conclude/resume the discussion.

**Reply by adlrocha** [reply]
I agree with @vukolic. Our goal is performance and fast finality in subnets, so I would go with option 2 (maybe considering modularity in the process, thinking about the larger set of common abstractions that we can use for all consensus algorithms so we don't have to re-write several of the layers for every new consensus).

Also, considering recent developments, slightly departing from the Filecoin ecosystem as long as we target high performance shouldn't be a concern.

**Comment by LefKok** [comment]
Bringing back attention to the discussion on chain-quality and Narwhal/ISS (https://github.com/protocol/ConsensusLab/discussions/91), TL;DR, the parallelisation benefit comes at the cost of higher latencies if a client is attached to a malicious validator

**Comment by aakoshh** [comment]
https://github.com/aakoshh/forest/pull/6 lists some vulnerabilities in _italics_ that come from not signing blocks in Option 2.
